### PR TITLE
coz: Causal profiler

### PR DESCRIPTION
**Pull request by klimkin**
Coz is a new kind of profiler that unlocks optimization opportunities
missed by traditional profilers. Coz employs a novel technique we call
causal profiling that measures optimization potential. This measurement
matches developers' assumptions about profilers: that optimizing
highly-ranked code will have the greatest impact on performance. Causal
profiling measures optimization potential for serial, parallel, and
asynchronous programs without instrumentation of special handling for
library calls and concurrency primitives. Instead, a causal profiler
uses performance experiments to predict the effect of optimizations.
This allows the profiler to establish causality: "optimizing function X
will have effect Y," exactly the measurement developers had assumed they
were getting all along.

https://github.com/plasma-umass/coz

Specify library name and version:  **coz/0.2.2**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/coz/all/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+cmake_minimum_required(VERSION 3.1)
```
Comment (prince-chrismc) [inline recipes/coz/all/CMakeLists.txt:1]: :thinking: Why write an entire build scrip if upstream already has one?

You can use the Autotools for make file projects... this would eliminate us having to maintain this cmake in the future.

File: recipes/coz/all/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+file(GLOB_RECURSE headers source_subfolder/include/*.h)
+
+add_library(${PROJECT_NAME} SHARED ${sources})
+set_target_properties(${PROJECT_NAME} PROPERTIES
+    PUBLIC_HEADER "${headers}"
+    VERSION ${CONAN_PACKAGE_VERSION})
+target_include_directories(${PROJECT_NAME}
+    PUBLIC source_subfolder/libcoz
+    PRIVATE source_subfolder/include)
+target_link_libraries(${PROJECT_NAME} PUBLIC dl rt Threads::Threads elf++ dwarf++)
```
Comment (prince-chrismc) [inline recipes/coz/all/CMakeLists.txt:24]: Pthread and rt are missing for package info

Comment (prince-chrismc) [reply recipes/coz/all/CMakeLists.txt:24]: Should self and drawf be atted to the required? They are available in CCI as I recall

File: recipes/coz/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    exports_sources = "CMakeLists.txt", "patches/*"
+    generators = "cmake"
+
+    _cmake = None
+    _source_subfolder = "source_subfolder"
+
+    def configure(self):
+        if self.settings.os == "Macos" or self.settings.compiler == "Visual Studio":
+            raise ConanInvalidConfiguration(
+                "libelfin doesn't support compiler: {} on OS: {}.".format(
```
Comment (prince-chrismc) [inline recipes/coz/all/conanfile.py:27]: Wrong name

File: recipes/coz/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            tools.patch(**patch)
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "Coz"
```
Comment (prince-chrismc) [inline recipes/coz/all/conanfile.py:55]: Upstream does not have CMake I am curious where this name came from

File: recipes/coz/all/test_package/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package C)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} test_package.c)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
Comment (prince-chrismc) [inline recipes/coz/all/test_package/CMakeLists.txt:8]: This the parent requires c++11 you'll need to set the c property here 

On mobile but I assume this is the failure

File: recipes/coz/all/test_package/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+
+    def build(self):
+        cmake = CMake(self, build_type="RelWithDebInfo")
```
Comment (prince-chrismc) [inline recipes/coz/all/test_package/conanfile.py:10]: Build type?

File: recipes/coz/all/conanfile.py (hunk 4 above)
Comment (prince-chrismc) [reply recipes/coz/all/conanfile.py:55]: https://github.com/plasma-umass/coz/blob/master/coz-profilerConfig.cmake.in

**Comment by klimkin** [comment]
@prince-chrismc Thank you for prompt review. I will try to push CMake build upstream in the next two weeks.

**Comment by klimkin** [comment]
Making some progress here: https://github.com/plasma-umass/coz/pull/178

File: recipes/coz/all/CMakeLists.txt (hunk 1 above)
Comment (klimkin) [reply recipes/coz/all/CMakeLists.txt:1]: The upstream uses CMake to build now.

**Comment by prince-chrismc** [comment]
Missing binaries.

You'll need to open a new PR for that dep

**Comment by klimkin** [comment]
> Missing binaries.
> 
> You'll need to open a new PR for that dep

Which dep? libelfin was there for a long time. If it was missing, other builds would fail too, but many passed.

**Comment by prince-chrismc** [comment]
Ah on mobile... there was a change to CI less then 3 weeks which added new compilers ... it's in one of the docs pages AFAIK

**Comment by klimkin** [comment]
@prince-chrismc I am lost. Couldn't find anything in CCI docs about re-triggering the build. libelfin/0.3 was part of CCI for some time.

**Comment by prince-chrismc** [comment]
### TL;DR

> You'll need to open a new PR for that dep

Just change a white space if you see nothing to fix

***

Okay, let me take some time to give some context, you make a lot of great contributions and I think writing this down be good. (this is my understanding as a community member) 

1. CCI is in "Early Access Program"
2. CCI changes a LOT and very frequently
3. New rule get added to ensure the widest range of compatibility
3. Sometimes when changes (to the back-end infrastructure) new additional requirements are added
4. There is a "tapaholes job" to fill-in in the "new requirements"
5. This sometimes is blocked by the "new rules"
    * ie. sometimes the new rules prevent building the new requirements
6. The tapaholes job is very experimental and I am not sure it's running at 100%

Ultimately this means a human, you or I, will need to manually open a pull request.... If it passes the infra team might manually intervene and prompt/upload them to bintray.


There's a lot of complicated moving parts that need to be combined and it's not trivial.


Hopefully I've helped explain why there's a problem (the action item is on the top of the post)

If you have anymore questions, I'll do my best to answer!

**Comment by klimkin** [comment]
Got it, thanks!

**Comment by klimkin** [comment]
Triggered libelfin/3.0 rebuild: https://github.com/conan-io/conan-center-index/pull/5069

**Comment by prince-chrismc** [comment]
You need to wait for the other one to be merged first

**Comment by prince-chrismc** [comment]
You were too fast! The curse should be broken I can install the binaries locally

File: recipes/coz/all/test_package/CMakeLists.txt (hunk 5 above)
Comment (danimtb) [reply recipes/coz/all/test_package/CMakeLists.txt:8]: @klimkin I believe this was not fixed

File: recipes/coz/all/test_package/conanfile.py (hunk 6 above)
Comment (danimtb) [reply recipes/coz/all/test_package/conanfile.py:10]: Why forcing built type in test package?

Comment (SSE4) [inline recipes/coz/all/test_package/conanfile.py:10]:
```suggestion
        cmake = CMake(self)
```

Comment (klimkin) [reply recipes/coz/all/test_package/conanfile.py:10]: To work properly Coz tool requires debug information. That's why the build forced to `RelWithDebInfo` in the test_package.

> To run your program with coz, you will need to build it with debug information. You do not need to include debug symbols in the main executable: coz uses the same procedure as gdb to locate debug information for stripped binaries.

https://github.com/plasma-umass/coz

Comment (klimkin) [reply recipes/coz/all/test_package/conanfile.py:10]: To work properly Coz tool requires debug information. That's why the build forced to `RelWithDebInfo` in the test_package.

> To run your program with coz, you will need to build it with debug information. You do not need to include debug symbols in the main executable: coz uses the same procedure as gdb to locate debug information for stripped binaries.

https://github.com/plasma-umass/coz

File: recipes/coz/all/conanfile.py (hunk 3 above)
Comment (klimkin) [reply recipes/coz/all/conanfile.py]: Fixed

File: recipes/coz/all/conanfile.py (hunk 4 above)
Comment (klimkin) [reply recipes/coz/all/conanfile.py:57]: Upstream now uses CMake

File: recipes/coz/all/test_package/conanfile.py (hunk 6 above)
Comment (prince-chrismc) [reply recipes/coz/all/test_package/conanfile.py:10]: Pretty please :pray:

```suggestion
        cmake = CMake(self, build_type="RelWithDebInfo") # To work properly Coz tool requires debug information https://github.com/plasma-umass/coz
```

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

File: recipes/coz/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    settings = "os", "arch", "compiler", "build_type"
+
+    requires = "libelfin/0.3"
+    exports_sources = "CMakeLists.txt", "patches/*"
+    generators = "cmake", "cmake_find_package"
+
+    _source_subfolder = "source_subfolder"
+
+    def configure(self):
```
Comment (uilianries) [inline recipes/coz/all/conanfile.py:24]:
```suggestion
    def validate(self):
```

File: recipes/coz/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            raise ConanInvalidConfiguration(
+                "coz doesn't support compiler: {} on OS: {}.".format(
+                    self.settings.compiler, self.settings.os))
+        if self.settings.compiler.cppstd:
+            tools.check_min_cppstd(self, "11")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = glob.glob(self.name + "-*")[0]
+        os.rename(extracted_dir, self._source_subfolder)
```
Comment (uilianries) [inline recipes/coz/all/conanfile.py:38]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version],
                  destination=self._source_subfolder, strip_root=True)
```

File: recipes/coz/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import glob
```
Comment (uilianries) [inline recipes/coz/all/conanfile.py:1]:
```suggestion
```

File: recipes/coz/all/test_package/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package C)
+
+set(CMAKE_CXX_STANDARD 11)
```
Comment (uilianries) [inline recipes/coz/all/test_package/CMakeLists.txt:4]:
```suggestion
```

C project.

File: recipes/coz/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    def build(self):
+        self._cmake.configure()
+        self._cmake.build()
+
+    def package(self):
+        self._cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "Coz"
```
Comment (uilianries) [inline recipes/coz/all/conanfile.py:57]:
```suggestion
        self.cpp_info.names["cmake_find_package"] = "Coz"
        self.cpp_info.names["cmake_find_package_multi"] = "Coz"
```
Need to apply same for multi, otherwise it will generate the wrong name

File: recipes/coz/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        self.cpp_info.syslibs = ["dl", "rt"]
+        if self.settings.os != "Windows":
+            self.cpp_info.system_libs = [
+                "-Wl,--push-state,--no-as-needed", "-ldl", "-Wl,--pop-state"
+            ]
```
Comment (uilianries) [inline recipes/coz/all/conanfile.py:62]:
```suggestion
        if self.settings.os == "Linux":
            self.cpp_info.system_libs = ["dl", "rt", "pthread"]
```

cpp_info.syslibs is illegal, doesn't exist.
Thread module is listed in upstream's cmake file

**Review by uilianries — CHANGES_REQUESTED** [review]

**Review by prince-chrismc — APPROVED** [review]

File: recipes/coz/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        self._cmake.configure()
+        self._cmake.build()
+
+    def package(self):
+        self._cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "Coz"
+        self.cpp_info.names["cmake_find_package_multi"] = "Coz"
```
Comment (SpaceIm) [inline recipes/coz/all/conanfile.py:56]:
```suggestion
        self.cpp_info.filenames["cmake_find_package"] = "coz-profiler"
        self.cpp_info.filenames["cmake_find_package_multi"] = "coz-profiler"
        self.cpp_info.names["cmake_find_package"] = "coz"
        self.cpp_info.names["cmake_find_package_multi"] = "coz"
```

File: recipes/coz/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs = ["dl", "rt", "pthread"]
```
Comment (SpaceIm) [inline recipes/coz/all/conanfile.py:58]: private dependencies, so probably useless because coz is always shared, but it doesn't hurt I guess

File: recipes/coz/all/conanfile.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
+import os
+from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
+
+
```
Comment (SpaceIm) [inline recipes/coz/all/conanfile.py:5]:
```suggestion
from conans.errors import ConanInvalidConfiguration

required_conan_version = ">=1.33.0"


```

File: recipes/coz/all/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.1)
+project(coz)
```
Comment (SpaceIm) [inline recipes/coz/all/CMakeLists.txt:2]:
```suggestion
project(cmake_wrapper)
```

**Comment by uilianries** [comment]
@klimkin please, consider SpaceIm's review, so we are good to go.

File: recipes/coz/all/conanfile.py (hunk 14 above)
Comment (klimkin) [reply recipes/coz/all/conanfile.py]: See https://github.com/conan-io/conan-center-index/pull/3794/commits/95b185257b53f5cd20557a072f8300e27f80956c

**Review by SpaceIm — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
